File: bin/book.erl
#!/usr/bin/env escript

-define(BOOKPATH, '/data/books').

list_files(RegExp) ->
    filelib:fold_files(?BOOKPATH, "((?i)" ++ RegExp ++ ")", true,
                       fun(File, Acc) -> [File|Acc] end, []).

% Returns [{Index, {File, Path}}]
process_paths(Paths) ->
    Sorted = lists:sort(fun({A,_}, {B,_}) -> A < B end,
                        [{filename:basename(P), P} || P <- Paths]),
    lists:zip(lists:seq(1, length(Sorted)), Sorted).

main(Args) ->
    Paths = process_paths(case length(Args) of
                              0 -> list_files("");
                              _ -> list_files(hd(Args))
                          end),
    case length(Paths) of
        0 ->
            io:format("Nothing found!~n");
        _ ->
            [io:format("\e[32m~3w \e[33m~s\n",
                       [Index, File]) || {Index, {File, _}} <- Paths],
            {ok, [Choice|_]} = io:fread("\e[32mOpen>\e[0m ", "~d"),
            {_, {_, Path}} = lists:nth(Choice, Paths),
            open_port({spawn_executable, "/usr/bin/open"}, [{args, [Path]}])
    end.

File: bin/book.rb
#!/usr/bin/env ruby

require 'find'

module Find
  def populate(*paths)
    matched = {}
    find(*paths) do |path|
      if File.file?(path) && yield(path)
        matched[File.basename(path, File.extname(path))] = path
      end
    end
    matched
  end
  module_function :populate
end

class Books
  def initialize(path, term = '')
    @books = find(path, term)
    if @books.empty?
      puts "No matches."
      return
    end
    select
  end

  def find(path, term)
    Find.populate(path) do |p|
      p =~ /#{term}/i
    end.sort
  end

  def select
    @books.each_with_index do |book, i|
      puts "\e[32m %-4d \e[33m%s" % [ i, book.flatten.first ]
    end

    print "\e[31m Select:\e[0m "
    begin
      choice = STDIN.gets
    rescue Interrupt
      # Catch CTRL-C.
    end
    return if choice.nil?
    choice = choice.chomp.to_i

    open @books[choice].flatten.last
  end

  def open(path)
    exec 'open', path
  end
end

Books.new('/data/books', ARGV.join(' '))
